compiler/src/iree/compiler/PluginAPI/PluginManager.cpp

```diff
@@ -0,0 +1,136 @@
+
+AbstractPluginRegistration::~AbstractPluginRegistration() = default;
+AbstractPluginSession::~AbstractPluginSession() = default;
+
+void PluginManagerOptions::bindOptions(OptionsBinder &binder) {
+  static llvm::cl::OptionCategory category("IREE compiler plugin options");
+
+  binder.list<std::string>("iree-plugin", plugins,
+                           llvm::cl::desc("Plugins to activate"),
+                           llvm::cl::cat(category));
+  binder.opt<bool>(
+      "iree-print-plugin-info", printPluginInfo,
+      llvm::cl::desc("Prints available and activated plugin info to stderr"),
+      llvm::cl::cat(category));
+}
+
+PluginManager::PluginManager() {}
+
+bool PluginManager::loadAvailablePlugins() {
+// Initialize static plugins.
+#define HANDLE_PLUGIN_ID(plugin_id) \
+  if (!iree_register_compiler_plugin_##plugin_id(this)) return false;
+#include "iree/compiler/PluginAPI/Config/StaticLinkedPlugins.inc"
+#undef HANDLE_PLUGIN_ID
+  return true;
+}
+
+void PluginManager::globalInitialize() {
+  for (auto &kv : registrations) {
+    kv.second->globalInitialize();
+  }
+}
+
+void PluginManager::initializeCLI() {
+  for (auto &kv : registrations) {
+    kv.second->initializeCLI();
+  }
+}
+
+void PluginManager::registerDialects(DialectRegistry &registry) {
+  for (auto &kv : registrations) {
+    kv.second->registerDialects(registry);
+  }
+}
+
+void PluginRegistrar::registerPlugin(
+    std::unique_ptr<AbstractPluginRegistration> registration) {
+  std::string_view id = registration->getPluginId();
+  auto foundIt = registrations.insert(
+      std::make_pair(llvm::StringRef(id), std::move(registration)));
+  if (!foundIt.second) {
+    llvm::errs() << "ERROR: Duplicate plugin registration for '" << id << "'\n";
+    abort();
+  }
+}
+
+LogicalResult PluginManagerSession::activatePlugins() {
+  auto getAvailableIds = [&]() -> llvm::SmallVector<llvm::StringRef> {
+    llvm::SmallVector<llvm::StringRef> availableIds;
+    for (auto &kv : pluginManager.registrations) {
+      availableIds.push_back(kv.first());
+    }
+    std::sort(availableIds.begin(), availableIds.end());
+    return availableIds;
+  };
+
+  // Print available plugins.
+  if (options.printPluginInfo) {
+    // Get the available plugins.
+    llvm::errs() << "[IREE plugins]: Available plugins: ";
+    llvm::interleaveComma(getAvailableIds(), llvm::errs());
+    llvm::errs() << "\n";
+  }
+
+  // Process activations.
+  // In the future, we may make this smarter by allowing dependencies and
+  // sorting accordingly. For now, what you say is what you get.
+  llvm::StringSet<> activatedPluginIds;
+  for (auto &pluginId : options.plugins) {
+    if (options.printPluginInfo) {
+      llvm::errs() << "[IREE plugins]: Activating plugin '" << pluginId
+                   << "'\n";
+    }
+    if (!activatedPluginIds.insert(pluginId).second) {
+      if (options.printPluginInfo) {
+        llvm::errs() << "[IREE plugins]: Skipping duplicate plugin '"
```

Comment:
No, it could come from --iree-plugin=example multiple times. I debated this (was originally failing), but decided it would be better to just ignore. Where were headed, there will end up being dependencies and layered command line flags... And those carping about duplicates seems like needless complexity.